### PR TITLE
LottieAnimation Extension

### PR DESCRIPTION
**Pull request by smfields**
This is an extension component that wraps the [Lottie Web](https://github.com/airbnb/lottie-web) Javascript library. The purpose of the library is to allow applications to directly render animations exported from Adobe After Effects without needing recreate the animation by hand in the HTML.

I've added the documentation page as well as a demo page were you can play around with the various animation controls:

### Docs Page
image

### Demo Page
image

This is my first contribution to this repo, so please let me know if I missed anything.

### CODE REVIEW

**Comment by smfields** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by stsrki** [comment]
Wow, I've just noticed this. There's a lot to review :) Thanks.

**Comment by stsrki** [comment]
Hi @smfields, can I please ask you to reformat the code to be along our [guidelines](https://github.com/Megabit/Blazorise/wiki/Contributing#conventions--guidelines) before I start with the review? Especially the alignments of `=`.

**Comment by smfields** [comment]
Hey @stsrki, hopefully it should all be good now.

**Comment by stsrki** [comment]
Much better, thanks. I can proceed from here :)

File: Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    animation.setLoop = function (loop) {
+        this.loop = loop;
+    }
+
+    animation.setSendCurrentFrame(options.sendCurrentFrame);
+    registerEvents(dotNetAdapter, animation);
+
+    return animation;
+}
+
```
Comment (stsrki) [inline Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js:41]: Can you also add a `destroy` method that will destroy the animation and remove any event listeners?

File: Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+async function invokeDotNetMethodAsync(dotNetAdapter, methodName, ...args) {
```
Comment (stsrki) [inline Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js:42]: We're trying to target the ES6 compatibility level. Can you make sure async/await is OK here?

Comment (smfields) [reply Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js:42]: Looks like async/await wasn't added until ECMAScript 2017, so I've swapped it out for the Promise APIs.

File: Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js (hunk 1 above)
Comment (smfields) [reply Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js:41]: There's already a `destroy` method the animations themselves that we call directly by returning a reference to the animation. 

That's what gets called in the [DisposeAnimation](https://github.com/Megabit/Blazorise/blob/905bb7b21526a85a5858b0cc9c5d537104eea937/Source/Extensions/Blazorise.LottieAnimation/LottieAnimation.razor.cs#L142) method, which is part of the component's `DisposeAsync` method.

Does that cover what we need?

Comment (stsrki) [reply Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js:41]: I've just now seen that you return `animation` object from JS and then store it in IJSObjectReference. We usually `export function destroy` and then call that one instead. And in IJSObjectReference is stored the JS module. I think it is better to stick to the same principle.

File: Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js (hunk 2 above)
Comment (stsrki) [reply Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js]: Looking good now, thanks.

File: Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js (hunk 1 above)
Comment (stsrki) [reply Source/Extensions/Blazorise.LottieAnimation/wwwroot/lottie-animation.js:41]: Now that I think, you can leave it as it is. It's a good approach, making it easier to work with the `animation` object and call its methods.

**Comment by stsrki** [comment]
@David-Moreira can you give a final comment on this PR? I think it is about right, can't see any issue left.

**Comment by David-Moreira** [comment]
I'll see If I can over this on the weekend.

**Review by David-Moreira — APPROVED** [review]
LGTM. Works great!

On the demo, If I remove `Style="@("height: 250px; background-color:" + _backgroundColor)"` it does not show at all. 
Can we have default dimensions? I can see users trying to use it for the first time, and easily forgetting that they configure dimensions.

File: Source/Extensions/Blazorise.LottieAnimation/Configuration/LoopConfiguration.cs [hunk 3]
```diff
@@ -0,0 +1,58 @@
+using System.Text.Json.Serialization;
+
+namespace Blazorise.LottieAnimation;
+
+/// <summary>
+/// Describes how an animation should loop.
+/// </summary>
+[JsonConverter( typeof( LoopConfigurationConverter ) )]
+public record LoopConfiguration
```
[the remark this answers is not in the document]
Comment (stsrki) [reply Source/Extensions/Blazorise.LottieAnimation/Configuration/LoopConfiguration.cs:9]: I'm not sure what you mean by 

> I remember you said there was some limitation with static types right

but we already have a similar approach to `LoopConfiguration`. Charts come to mind with their converters for indexable types.

File: Source/Extensions/Blazorise.LottieAnimation/LottieAnimation.razor.cs [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+                    if ( frameSyncRequired )
+                    {
+                        await SynchronizeCurrentFrame();
+                    }
+                    else if ( pausedChanged )
+                    {
+                        // Synchronizing the current frame will already synchronize the pause setting, so we only need
+                        // to do it manually if we're not synchronizing the frame
+                        await SynchronizePaused();
+                    }
```
[the remark this answers is not in the document]
Comment (stsrki) [reply Source/Extensions/Blazorise.LottieAnimation/LottieAnimation.razor.cs]: Added `Task.WhenAll` to execute multiple tasks.

**Comment by stsrki** [comment]
> On the demo, If I remove `Style="@("height: 250px; background-color:" + _backgroundColor)"` it does not show at all.

I have tested it without a fixed height, and it works. The only downside when height is undefined is that the loading animation will not be visible. That's why I have added a **Best Practices** note for the fixed height.

**Review by stsrki — APPROVED** [review]

File: Source/Extensions/Blazorise.LottieAnimation/LottieAnimation.razor.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (David-Moreira) [reply Source/Extensions/Blazorise.LottieAnimation/LottieAnimation.razor.cs]: It still has the js interop cost for each call. And I have no idea if lottie will handle concurrency well... 
Let's hope it does. :)

[the remark this answers is not in the document]
Comment (stsrki) [reply Source/Extensions/Blazorise.LottieAnimation/LottieAnimation.razor.cs]: Ah, you meant to send all to JS in one call. I don't think we will have that many benefits from one-call, considering the types of operations involved.

**Comment by stsrki** [comment]
@smfields I failed to say thank you. You did a great job with this PR. Consider yourself as an [MVC candidate](https://github.com/Megabit/Blazorise/wiki/What-it-takes-to-be-a-Blazorise-MVC) for the following year. Thank you!!

**Pull request merged**